### DISCUSSION TITLE
Config environment variables?

### DISCUSSION THREAD

**Discussion by epage** [GitoxideLabs/gitoxide#147] [Ideas]
As of [git 2.31](https://github.com/git/git/blob/311531c9de557d25ac087c1637818bd2aad6eb3a/Documentation/RelNotes/2.31.0.txt#L127-L130)
```
 * Two new ways to feed configuration variable-value pairs via
   environment variables have been introduced, and the way
   GIT_CONFIG_PARAMETERS encodes variable/value pairs has been tweaked
   to make it more robust.
```

Unsure if compatibility with this is within scope for gitoxide but I thought I'd share a crate I just made that might be useful, [git-config-env](https://github.com/crate-ci/git-config-env)

`git-config-env` supports:
- `GIT_CONFIG_PARAMETERS ` env var is for passing `git -c key=value` from the `git` process to an external sub-command.  There are two ways of quoting the value being passed down, pre-2.31 and 2.31+.  Like with git, git-config-env's parser supports reading both.
- `GIT_CONFIG_COUNT`, `GIT_CONFIG_KEY_<n>`, `GIT_CONFIG_VALUE_<n>` env vars are a way for anyone to pass configuration through the environment.
- Parsing the `-c` argument

Current limitations:
- Does not parse keys
- UTF-8 only
- Ignores errors (non-UTF-8, something missing)
- While the happy path (no escaping) is kept to the minimum allocations, I was lazy in writing the escaping code path for `GIT_CONFIG_PARAMETERS ` which does some allocations to simplify parsing bookkeeping
- Does not support generating `GIT_CONFIG_PARAMETERS `

These were out of scope for my needs but can definitely add or accept contributions if these are needed by gitoxide.

Note: doc.rs' queue is backed up so as of writing this, docs are not posted yet.

**Comment by edward-shen** [comment]
Heyo, maintainer of `git-config` here!

`git-config` is currently in a state where it only supports reading and writing for a single file—it doesn't have any support for collecting configs across the system platform (and environment for that matter) to present a unified config.

It's definitely something I think would be useful in the future, just not right now! `git-config` just needs a little more maturation (which, admittedly, I've been putting of for other projects) before we can start taking use of this.

I haven't taken an in-depth look, but if the time comes, how would you like `git-config` to use your code? Merged into `git-config`, or as a dependency?

**Reply by epage** [reply]
> I haven't taken an in-depth look, but if the time comes, how would you like git-config to use your code? Merged into git-config, or as a dependency?

I think we'll need to play it by ear.  My original motivation for `git-config-env` was to help in writing an external sub-command. Pulling in all of `git-config` depends on
- The size of it (and dependencies) at the time this happens
- If mine (and any other) external subcommands would find it worth migrating from `git2::Config` to `git-config` such we'd already be pulling it in anyways.
